README.md
# Mapeo Server

Mapeo server, used for managing observation data. It includes observation and
media management routes and static file routes for an offline tile server.

## Install

```
$ npm install mapeo-server
```

## API

> var MapeoServer = require('mapeo-server')

### var server = MapeoServer(osm, media[, opts])

Creates a mapeo http server.

`osm` is a [kappa-osm][kappa-osm] instance.

`media` is an instance of [abstract-blob-store][abstract-blob-store]. In order to do sync correctly, it must be patched with a `list(cb)` method, or you can use a blob store that already supports this, like [safe-fs-blob-store][safe-fs-blob-store].

Valid `opts` include
- `staticRoot` (string): the filesystem path to serve static files, like styles
  and tiles from.
- `fallbackPresetsDir` (string): the filesystem path to serve as a fallback for
  requests to `/presets` if no files exist in `staticRoot/presets`. Useful for
  serving defaults that can be overridden by user files in `staticRoot`.

### var handled = server.handle(req, res)

Tries to handle an http request `req`, writing the result to `res`. Returns true if it was handled, and false if not.

## Routes

The following routes are available.

### Device

#### `GET /device/id`

Retrieve this device's unique 64-character ID, as a JSON string. E.g.

```json
"e685da71b0c0ecc19feac1a419b1064889cde89496ad58e973daf4b323edd471"
```

### Observations

#### `GET /observations?bbox=a,b,c,d`

Get list of observations. Currently, `bbox` is ignored and all observations are
returned. The response is a JSON array of observation objects. E.g.

```json
[
  { "type": "observation", "lat": 12.1245, "lon": -0.3243 },
  ...
]
```

#### `POST /observations`

Creates an observation. Expects a single JSON object representing the
observation. The following fields are required:

- `type` must be `observation`

The object will be returned, with the fields `id`, `version`, `created_at` and `timestamp` set.

The property `ref` can also be optionally set, to indicate that the observation *observes* the OSM element with ID `ref`.

#### `GET /observations/:id`

Fetch an observation by its `id`. An array of JSON objects will be returned.
Usually there will only be one result, but in forking situations (e.g. two
devices create offline edits of the same observation then sync) there can be
multiple results.

#### `PUT /observations/:id`

Update an observation by its `id` by providing a JSON object representing the new observation. `id` and `version` *must* be set. Only the following fields can be modified:

- `lat`
- `lon`
- `ref`
- `attachments`
- `tags`
- `metadata`
- `schemaVersion`

#### `PUT /observations/to-element/:id`

Converts an observation, by its ID, to an OSM element (defaults to `node`
currently). Returns an object of the form `{ id: elementId }`. If the
observation has already been converted to a node, it will return the existing
ID.

### Presets

#### `GET /presets`

Returns a JSON array with the names of available presets. E.g.

```json
[
  "default_osm",
  "jungle",
  "waorani"
]
```

#### `GET /presets/:id/*`

Fetch a static file belonging to a preset with id `id`.

### Media

#### `POST /media`

Save a photo to the database, expects a JSON object in the body of the request
with the following properties:

- `original` (string)
- `preview` (string)
- `thumbnail` (string)

Each of these properties must be an absolute filepath pointing to the location
of the media in the local filesystem. The files will be copied to the database
and can be safely deleted once the request completes successfully. All 3 sizes
are required and should be the following sizes:

- *original*: full-size original media
- *preview*: maximum dimension 2000px
- *thumbnail*: maximum dimension 400px

Original media is not syncronized between mobile devices, but previews and
thumbnails are. Original media is always syncronized with desktop devices. In
the future Mapeo may more intelligently manage full-size media in order to save
space on mobile.

A single JSON object is returned, with the media's unique `id`:

```json
{
  "id": "225961fb85d82312e8c0ed511.jpg",
}
```

To fetch the thumbnail later, one would hit the route `GET /media/thumbnail/225961fb85d82312e8c0ed511.jpg`.

#### `GET /media/:type/:id`

Retrieve a piece of media (photos only for now) by its `id`. Valid `type`s are
`original`, `preview` and `thumbnail`. e.g.

```
GET /media/thumbnail/225961fb85d82312e8c0ed511.jpg
```
or
```
GET /media/original/225961fb85d82312e8c0ed511.jpg
```

### Mapbox Styles & Tiles

#### `GET /styles`

Returns a JSON array with the names of all available vector tilesets. E.g.

```js
[
  {
    id: 'satellite-v9',
    name: 'Satellite',
    bounds: [ -122.339973, 37.742214, -122.150116, 37.856694 ],
    minzoom: 0, maxzoom: 22
  }
]
```

#### `GET /styles/:id/style.json`

Retrieve the `style.json` file for a given style.

#### `GET /styles/:id/tiles/:x/:y/:z.:ext`

Fetch a single vector tile from the tileset `id` by an `x`,`y`,`z` coordinate.

#### `GET /styles/:id/tiles/:tileid/:x/:y/:z.:ext`

Fetch a single vector tile from the tileset `id` by an `x`,`y`,`z` coordinate.
Supports a `tileid` which would be a subdirectory or an asar file under the
`tiles/` directory, so support multiple tilesets per dataset id.

### Sync

#### `GET /sync/listen`

Begin the sync server for listening for connections.  Returns 200 OK once server is up and running.

#### `GET /sync/join`

Join the network and become discoverable by other peers.

Optionally, a url parameter `name` for this peer can be provided, to appear in
others' `/sync/targets` list.

Pass the url parameter `project_key` to only discover other peers also interested in the same project. `project_key` should be a hex-encoded random 32-bytes, e.g. `crypto.randomBytes(32).toString('hex')`

#### `GET /sync/leave`

Leave the network and no longer be discoverable.

Pass the url parameter `project_key` if you passed it into `GET /sync/join` to unsubscribe to this particular project.

#### `GET /sync/destroy`

Destroy the current sync server and close all open connections. Returns 200 OK once connections have terminated.

#### `GET /sync/peers`


Options

  * `interval`: Default false. If set, will send an updated list of the peers at the
    given interval using streaming GET request.

Returns list of available sync peers every set interval. Only lists other devices broadcasting using the 'mapeo-sync' key.

Each sync target is an object with `ip`, `port`, `host`, and `name` if device
name is set.

```js
GET /sync/peers?interval=300

// then every 300ms...
{ topic: 'peers', message: [{ip: '127.0.0.1', port: 1337, host: 'hostname', name: "Karissa's Device", {state: StateObject}}]}
{ topic: 'peers', message: [{name: "/path/to/my/syncfile.mapeodata", {state: StateObject}}]}

```

A `StateObject` has a `topic` and `message` which give you an idea of what the current state

#### `GET /sync/start`

Options

  * `filename`: For local filesystem sync, provide filename
  * `port` and `host`: To sync with another target through TCP (UDP?)

Start syncing and listen to progress events. Events are returned as a newline-delimited JSON stream.

Events are returned with a `topic` and `message` key:

```js
{"topic": "replication-started" }
{"topic": "replication-progress", {"db": { "sofar": 5, "total": 10 }, "media": { "sofar": 12, "total": 95 } } }
{"topic": "replication-error", "message": "Some error message here"}
{"topic": "replication-complete" }

Valid event topics:

  * `replication-error`: Sent once there is error, and the stream is closed.
  * `replication-started`: Sent once to indicate replication has started, but no data has been sent.
  * `replication-progress`: Sent for each datum (map element, photo) sent.
  * `replication-complete`: Sent once for a replication success, and the stream is closed.

Example client code for `/sync/start`
```js
var hyperquest = require('hyperquest')
var target = {filename: '/path/to/my/database.mapeodata'}
var url = `http://${host}/sync/start?${querystring.stringify(target)}`
var hq = hyperquest(url)
var stream = pump(hq, split2())
stream.on('data', function (data) {
  var row = JSON.parse(data)
  if (row.topic === 'replication-progress') console.log('progress...', row.message)
  if (row.topic === 'replication-error') console.log('error', row.message)
  if (row.topic === 'replication-complete') console.log('done')
})

```

## Usage

```js
var osmdb = require('kappa-osm')
var kappa = require('kappa-core')
var raf = require('random-access-file')
var level = require('level')
var blobstore = require('safe-fs-blob-store')
var Router = require('mapeo-server')

var osm = osmdb({
  core: kappa('./db', {valueEncoding: 'json'}),
  index: level('./index')),
  storage: function (name, cb) {
    process.nextTick(cb, null, raf(path.join(dir, 'storage', name)))
  }
})
var media = blobstore('./media')

var root = '/path/to/my/static/files' // optional
var route = Router(osm, media, {staticRoot: root})

var http = require('http')
var server = http.createServer(function (req, res) {
  var fn = route.handle(req, res)
  if (fn) {
    fn()
  } else {
    res.statusCode = 404
    res.end('not found\n')
  }
})
server.listen(5000)
server.on('close', function () {
  route.api.close()
})
```

### Use as Express middleware

```js
app.use('/api', Router(dir))
```

## Caveats

Relies on `fs` for media copying right now: this is a bug.

[kappa-osm]: https://github.com/digidem/kappa-osm
[abstact-blob-store]: https://github.com/maxogden/abstract-blob-store
[safe-fs-blob-store]: https://github.com/noffle/safe-fs-blob-store

## Community

Connect with the Mapeo community for support & to contribute!

- [**Mailing List**](https://lists.riseup.net/www/info/mapeo-en) (English)
- [**Mailing List**](https://lists.riseup.net/www/info/mapeo-es) (Spanish)
- [**IRC**](https://kiwiirc.com/nextclient/irc.freenode.net/) (channel #ddem)
- [**Slack**](http://slack.digital-democracy.org)

## License

MIT


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [17.1.2](https://github.com/digidem/mapeo-server/compare/v17.1.1...v17.1.2) (2021-11-16)

### [17.1.1](https://github.com/digidem/mapeo-server/compare/v17.1.0...v17.1.1) (2021-10-12)


### Bug Fixes

* Upgrade mapeo core for sync file fix ([190a621](https://github.com/digidem/mapeo-server/commit/190a621ea13723ac02e386ec829a7587953ca146))

## [17.1.0](https://github.com/digidem/mapeo-server/compare/v17.0.7...v17.1.0) (2021-02-05)


### Features

* Add websocket sync via mapeo-core@8.5.0 ([#61](https://github.com/digidem/mapeo-server/issues/61)) ([2441799](https://github.com/digidem/mapeo-server/commit/244179975d1769b6d59f5d5be1730128d489d31b))

### [17.0.7](https://github.com/digidem/mapeo-server/compare/v17.0.6...v17.0.7) (2020-12-21)

### [17.0.1](https://github.com/digidem/mapeo-server/compare/v17.0.0...v17.0.1) (2019-11-25)

## [17.0.0](https://github.com/digidem/mapeo-server/compare/v16.1.1...v17.0.0) (2019-11-14)


### ⚠ BREAKING CHANGES

* Use 32-byte project_key as url param, instead of project_id
* Fix sync issues with update to mapeo-core 8 and kappa-core 4

### Features

* Use 32-byte project_key as url param, instead of project_id ([fff3226](https://github.com/digidem/mapeo-server/commit/fff322623f33aa1507acd546b73a27b2cb67bec7))


### Bug Fixes

* Fix sync issues with update to mapeo-core 8 and kappa-core 4 ([d2aa43e](https://github.com/digidem/mapeo-server/commit/d2aa43ecab9f0dcf4e8016aebcecb60206e93744))
* Fix tests which expected device_id ([2328e6d](https://github.com/digidem/mapeo-server/commit/2328e6d271013d1bcd02fe3f997fd9c19a218f2e))

## [16.1.1](https://github.com/digidem/mapeo-server/compare/v16.1.0...v16.1.1) (2019-09-17)


### Bug Fixes

* Pass missing query screen to syncLeave ([8878f37](https://github.com/digidem/mapeo-server/commit/8878f37))



## [16.1.0](https://github.com/digidem/mapeo-server/compare/v16.0.1...v16.1.0) (2019-09-12)


### Features

* Allow a project_id to be specified for swarming ([#51](https://github.com/digidem/mapeo-server/issues/51)) ([d7eb75f](https://github.com/digidem/mapeo-server/commit/d7eb75f))

* Add /device/id route ([#50](https://github.com/digidem/mapeo-server/issues/50)) ([b582ef9](https://github.com/digidem/mapeo-server/commit/b582ef9))

# [15.0.0](https://github.com/digidem/mapeo-server/compare/v14.0.0...v15.0.0) (2019-05-04)


### Bug Fixes

* Correct cache-control and content-type headers ([bcf2bd3](https://github.com/digidem/mapeo-server/commit/bcf2bd3))


### Features

* **media:** Add preview images as well as thumbnail and original [BREAKING] ([#47](https://github.com/digidem/mapeo-server/issues/47)) ([0189c29](https://github.com/digidem/mapeo-server/commit/0189c29))


### BREAKING CHANGES

* **media:** Changes media route to POST instead of PUT, because PUT should be id-empotent, whereas this endpoint creates new resources for every POST.
* **media:** Pass parameters in the body of the request, rather than query string, to avoid url encoding issues
* **media:** Pass original as `original` property, rather than `file`
* **media:** All 3 image sizes are required


test/sync.js
var test = require('tape')
var needle = require('needle')
var pump = require('pump')
var hyperquest = require('hyperquest')
var through = require('through2')
var crypto = require('crypto')

var {
  listen,
  join,
  destroy,
  createServer,
  twoServers
} = require('./server')

test('sync - announce and close', function (t) {
  createServer(function (server, base) {
    needle.get(base + '/sync/listen', function (err, resp, body) {
      t.error(err)
      needle.get(base + '/sync/destroy', function (err, resp, body) {
        t.error(err)
        server.close()
        t.end()
      })
    })
  })
})

test('sync - two-server listen and dont find eachother', function (t) {
  twoServers(function (a, b) {
    listen(a, b, function (err) {
      t.error(err)
      var opts = {
        a: { projectKey: crypto.randomBytes(32).toString('hex') },
        b: { projectKey: crypto.randomBytes(32).toString('hex') }
      }
      join(a, b, opts, function (err) {
        t.error(err)
        var href = a.base + '/sync/peers?interval=100'
        var r = hyperquest(href, { end: false })
        var times = 0
        r.pipe(through.obj(function (data, enc, next) {
          var body = JSON.parse(data)
          t.equal(body.topic, 'peers')
          t.equal(body.message.length, 0)
          times += 1
          next()
        }))
        setTimeout(function () {
          r.destroy()
        }, 2000)
        r.on('error', function (err) {
          t.error(err)
        })
        r.on('end', function () {
          t.ok(times >= 2)
          destroy(a, b, function (err) {
            t.error(err)
            a.server.close()
            b.server.close()
            t.end()
          })
        })
      })
    })
  })
})

test('sync - two-server listen and join and find eachother', function (t) {
  twoServers(function (a, b) {
    a.router.api.core.sync.on('peer', function () {
      var href = a.base + '/sync/peers'
      var hq = hyperquest(href, { end: false })
      hq.pipe(through.obj(function (data, enc, next) {
        var body
        try {
          body = JSON.parse(data.toString())
        } catch (err) {
          t.fail('JSON parse failed: ' + data.toString())
        }
        t.equal(body.message.length, 1, 'message expected length')
        var entry = body.message[0]
        t.equal(entry.type, 'wifi')
        next()
      }))
      setTimeout(function () {
        hq.destroy()
      }, 250)
      hq.on('error', function (err) { t.error(err) })
      hq.on('end', function () {
        destroy(a, b, function (err) {
          t.error(err, 'destroyed without error')
          a.server.close()
          b.server.close()
          t.end()
        })
      })
    })
    listen(a, b, function (err) {
      t.error(err, 'server listening without error')
      join(a, b, function (err) {
        t.error(err, 'join without error')
      })
    })
  })
})

test('sync - two-server listen and join and find each other /w same project_key', function (t) {
  twoServers(function (a, b) {
    a.router.api.core.sync.on('peer', function () {
      var href = a.base + '/sync/peers'
      var hq = hyperquest(href, { end: false })
      hq.pipe(through.obj(function (data, enc, next) {
        var body
        try {
          body = JSON.parse(data.toString())
        } catch (err) {
          t.fail('JSON parse failed: ' + data.toString())
        }
        t.equal(body.message.length, 1, 'message expected length')
        var entry = body.message[0]
        t.equal(entry.type, 'wifi')
        next()
      }))
      setTimeout(function () {
        hq.destroy()
      }, 250)
      hq.on('error', function (err) { t.error(err) })
      hq.on('end', function () {
        destroy(a, b, function (err) {
          t.error(err, 'destroyed without error')
          a.server.close()
          b.server.close()
          t.end()
        })
      })
    })
    const projectKey = crypto.randomBytes(32).toString('hex')
    listen(a, b, function (err) {
      t.error(err, 'server listening without error')
      var opts = {
        a: { projectKey: projectKey },
        b: { projectKey: projectKey }
      }
      join(a, b, opts, function (err) {
        t.error(err, 'join without error')
      })
    })
  })
})

test('sync - two-server sync', function (t) {
  twoServers(function (a, b) {
    function done () {
      destroy(a, b, function (err) {
        t.error(err)
        a.server.close()
        b.server.close()
        t.end()
      })
    }
    a.router.api.core.sync.on('peer', function () {
      needle.get(a.base + '/sync/peers', function (err, resp, body) {
        t.error(err)
        t.equal(body.topic, 'peers')
        var entry = body.message[0]
        t.equal(body.message.length, 1)
        t.equal(entry.type, 'wifi')
        t.equal(entry.name, 'peer2')
        var href = a.base + `/sync/start?host=${entry.host}&port=${entry.port}`
        var hq = hyperquest(href, { end: false })
        var go = through.obj(function (data, enc, next) {
          var text
          try {
            text = JSON.parse(data.toString())
          } catch (err) {
            t.fail('JSON parse failed: ' + data.toString())
          }
          t.ok(RegExp(/replication-(complete|progress|started)/).test(text.topic))
          if (text.topic === 'replication-complete') {
            needle.get(a.base + '/sync/peers', function (err, resp, body) {
              t.error(err)
              var peers = body.message
              t.same(peers.length, 1, 'one peer')
              t.same(peers[0].state.topic, 'replication-complete')
              next()
            })
          } else next()
        })
        pump(hq, go, done)
      })
    })
    a._name = 'peer1'
    b._name = 'peer2'
    listen(a, b, function (err) {
      t.error(err)
      join(a, b, function (err) {
        t.error(err)
      })
    })
  })
})


test/tiles.js
var test = require('tape')
var hyperquest = require('hyperquest')
var createServer = require('./server')
var concat = require('concat-stream')

test('tiles: list', function (t) {
  createServer(function (server, base) {
    var href = base + '/styles'

    var hq = hyperquest.get(href, {})

    // http response
    hq.once('response', function (res) {
      t.equal(res.statusCode, 200, 'create 200 ok')
      t.equal(res.headers['content-type'], 'application/json', 'type correct')
    })

    // response content
    hq.pipe(concat({ encoding: 'string' }, function (body) {
      try {
        var obj = JSON.parse(body)
        var expected = [
          { id: 'satellite-v9',
            name: 'Satellite',
            bounds: [ -122.339973, 37.742214, -122.150116, 37.856694 ],
            minzoom: 0, maxzoom: 22
          },
          { id: 'satellite-streets-v9',
            name: 'Mapbox Satellite Streets',
            bounds: [ -122.339973, 37.742214, -122.150116, 37.856694 ],
            minzoom: 0, maxzoom: 16
          }
        ]
        t.deepEquals(obj, expected)
      } catch (e) {
        t.error(e, 'json parsing exception!')
      }
      server.close()
      t.end()
    }))
  })
})

test('tiles: get', function (t) {
  createServer(function (server, base) {
    var href = base + '/styles/sat-style/tiles/mapbox.satellite/6/10/24.png'
    var hq = hyperquest.get(href)
    hq.once('response', function (res) {
      t.equal(res.statusCode, 200, 'get 200 ok')
      t.equal(res.headers['content-type'], 'image/png', 'type correct')

      hq.pipe(concat(function (body) {
        t.equals(body.length, 21014, 'correct file length')
        server.close()
        t.end()
      }))
    })
    hq.once('error', function (err) {
      t.error(err, 'no http error')
    })
  })
})


index.js
var Router = require('routes')
var url = require('url')
var querystring = require('querystring')

var Api = require('./api')

module.exports = function (media, opts) {
  var router = Router()
  var api = Api(media, opts)

  // Presets
  router.addRoute('GET /presets', api.presetsList.bind(api))
  router.addRoute('GET /presets/:id/*', api.presetsGet.bind(api))

  // Media
  router.addRoute('GET /media/:type/:id', api.mediaGet.bind(api))
  router.addRoute('POST /media', api.mediaPost.bind(api))

  // Styles 'n Tiles
  router.addRoute('GET /styles', api.stylesList.bind(api))
  router.addRoute('GET /styles/:id/tiles/:tileid/:z/:y/:x.:ext',
    api.stylesGet.bind(api))
  router.addRoute('GET /styles/:id/tiles/:tileid/:z/:y/:x',
    api.stylesGet.bind(api))

  // Here, id is assumed to be the same as tileid
  router.addRoute('GET /styles/:id/tiles/:z/:y/:x.:ext',
    api.stylesGet.bind(api))
  router.addRoute('GET /styles/:id/tiles/:z/:y/:x',
    api.stylesGet.bind(api))

  router.addRoute('GET /styles/:id/style.json', api.stylesGetStyle.bind(api))
  router.addRoute('GET /styles/:id/*', api.stylesGetStatic.bind(api))

  return {
    api,
    handle: function (req, res) {
      var parsed = url.parse(req.url) // eslint-disable-line node/no-deprecated-api
      var q = querystring.parse(parsed.query)
      var m = router.match(req.method + ' ' + parsed.pathname)
      if (m) {
        m.fn(req, res, m.params, q)
        return true
      } else {
        return false
      }
    }
  }
}


api.js
var error = require('debug')('mapeo-server:error')
var fs = require('fs')
var path = require('path')
var body = require('body/json')
var randombytes = require('randombytes')
var asar = require('asar')
var ecstatic = require('ecstatic')
var Core = require('@mapeo/core')
var debounce = require('debounce')
var mime = require('mime')
var errors = Core.errors

module.exports = Api

const customMimeTypes = {
  'application/x-protobuf': ['pbf']
}
mime.define(customMimeTypes)

function Api (osm, media, opts) {
  if (!(this instanceof Api)) return new Api(osm, media, opts)
  if (!opts) opts = {}
  var defaultOpts = {
    id: 'MapeoDesktop_' + randombytes(8).toString('hex'),
    staticRoot: '.'
  }

  this.opts = Object.assign(defaultOpts, opts)
  this.staticRoot = this.opts.staticRoot
  this.ecstatic = ecstatic({
    cache: 60 * 5, // 5 minutes
    mimeTypes: customMimeTypes,
    root: this.staticRoot,
    handleError: false
  })
  if (opts.fallbackPresetsDir) {
    this.ecstaticFallbackPresets = ecstatic({
      cache: 60 * 5, // 5 minutes
      mimeTypes: customMimeTypes,
      root: opts.fallbackPresetsDir,
      baseDir: 'presets',
      handleError: false
    })
  }
  this.core = new Core(osm, media, this.opts)
  this.core.on('error', function (err) {
    error(err)
  })
}

function handleError (res, err) {
  if (typeof err === 'string') err = new Error(err)
  if (!err.status) err = errors(err)
  errors.send(res, err)
  error(err)
}

// Device
Api.prototype.deviceId = function (req, res) {
  this.core.getDeviceId(function (err, id) {
    if (err) return handleError(res, err)
    res.setHeader('content-type', 'application/json')
    res.end(JSON.stringify(id))
  })
}

// Observations
Api.prototype.observationDelete = function (req, res, m) {
  var self = this
  self.core.observationDelete(m.id, function (err) {
    if (err) return handleError(res, err)
    res.setHeader('content-type', 'application/json')
    res.end(JSON.stringify({ deleted: true }))
  })
}

Api.prototype.observationList = function (req, res, m, q) {
  this.core.observationList(q, function (err, results) {
    if (err) return handleError(res, errors(err))
    res.setHeader('content-type', 'application/json')
    res.end(JSON.stringify(results))
  })
}

Api.prototype.observationGet = function (req, res, m) {
  this.core.observationGet(m.id, function (err, obses) {
    if (err) return handleError(res, err)
    res.setHeader('content-type', 'application/json')
    res.end(JSON.stringify(obses))
  })
}

Api.prototype.observationCreate = function (req, res, m) {
  var self = this

  body(req, function (err, obs) {
    if (err) return handleError(res, errors.JSONParseError())

    self.core.observationCreate(obs, function (err, newObs) {
      if (err) return handleError(res, err)
      res.setHeader('content-type', 'application/json')
      res.end(JSON.stringify(newObs))
    })
  })
}

Api.prototype.observationUpdate = function (req, res, m) {
  var self = this

  body(req, function (err, newObs) {
    if (err) return handleError(res, errors.JSONParseError())

    if (newObs.id !== m.id) return handleError(res, errors.TypeMismatch(newObs.id, m.id))

    self.core.observationUpdate(newObs, function (err, finalObs) {
      if (err) return handleError(res, err)
      res.setHeader('content-type', 'application/json')
      res.end(JSON.stringify(finalObs))
    })
  })
}

Api.prototype.observationConvert = function (req, res, m) {
  this.core.observationConvert(m.id, function (err, id) {
    if (err) return handleError(res, err)
    res.setHeader('content-type', 'application/json')
    res.end(JSON.stringify({ id }))
  })
}

// Presets
Api.prototype.presetsList = function (req, res, m) {
  var target = path.join(this.staticRoot, 'presets')
  fs.readdir(target, function (err, files) {
    if (err) return handleError(res, err)
    files = files
      .filter(function (filename) {
        return fs.statSync(path.join(target, filename)).isDirectory()
      })
    res.setHeader('content-type', 'application/json')
    res.end(JSON.stringify(files))
  })
}

Api.prototype.presetsGet = function (req, res, m) {
  this.ecstatic(req, res, (err) => {
    if (err) return handleError(res, err)
    if (this.ecstaticFallbackPresets) {
      this.ecstaticFallbackPresets(req, res, (err) => {
        handleError(res, err || errors.NotFound())
      })
    } else {
      handleError(res, errors.NotFound())
    }
  })
}

// Media
Api.prototype.mediaGet = function (req, res, m) {
  var self = this
  var id = m.type + '/' + m.id

  this.core.media.exists(id, function (err, exists) {
    if (err) return handleError(res, err)
    if (!exists) return handleError(res, errors.NotFound())
    if (m.id.endsWith('.jpg')) res.setHeader('content-type', 'image/jpeg')
    else if (m.id.endsWith('.png')) res.setHeader('content-type', 'image/png')
    self.core.media.createReadStream(id).pipe(res)
  })
}

const expectedMediaFormats = ['original', 'preview', 'thumbnail']

Api.prototype.mediaPost = function (req, res, m, q) {
  const self = this
  let pending = expectedMediaFormats.length
  let errorSent = false

  body(req, function (err, media) {
    if (err) return handleError(res, errors.JSONParseError())
    if (!media) return handleError(res, new Error('Empty request body'))

    for (const format of expectedMediaFormats) {
      if (!media[format]) return error(400, new Error(`Request body is missing ${format} property`))
      if (!fs.existsSync(media[format])) return error(400, new Error(`File ${media[format]} does not exist`))
    }

    // Use the extension of the original media - assumes thumbnail and preview
    // is the same format / has the same extension.
    const ext = path.extname(media.original)
    const newMediaId = randombytes(16).toString('hex') + ext

    for (const format of expectedMediaFormats) {
      var destPath = format + '/' + newMediaId
      copyFileTo(media[format], destPath, done)
    }

    function done (err) {
      if (err) return error(500, new Error('There was a problem saving the media to the server'))
      if (--pending) return
      if (errorSent) return

      res.setHeader('content-type', 'application/json')
      res.end(JSON.stringify({ id: newMediaId }))
    }
  })

  function copyFileTo (file, to, cb) {
    var ws = self.core.media.createWriteStream(to, cb)
    fs.createReadStream(file).pipe(ws)
  }

  function error (code, error) {
    if (errorSent) return
    errorSent = true
    error.status = code
    handleError(res, error)
  }
}

// Tiles
Api.prototype.stylesList = function (req, res, m) {
  var self = this
  res.setHeader('content-type', 'application/json')
  fs.readdir(path.join(self.staticRoot, 'styles'), function (err, files) {
    if (err) return handleError(res, err)
    files = files
      .filter(function (file) {
        var stat = fs.statSync(path.join(self.staticRoot, 'styles', file))
        return stat.isDirectory() && fs.existsSync(path.join(self.staticRoot, 'styles', file, 'style.json'))
      })
      .map(function (dir) {
        var str = fs.readFileSync(path.join(self.staticRoot, 'styles', dir, 'style.json'), 'utf-8')
        if (str) {
          try {
            var json = JSON.parse(str)
            var srcTop = Object.keys(json.sources)[0] || {}
            var src = json.sources[srcTop]
            if (!src) return null
            return {
              id: dir,
              name: json.name,
              description: json.description,
              bounds: src.bounds,
              minzoom: src.minzoom,
              maxzoom: src.maxzoom
            }
          } catch (e) {
            return null
          }
        } else {
          return null
        }
      })
      .filter(Boolean)
    res.end(JSON.stringify(files))
  })
}

Api.prototype.stylesGetStyle = function (req, res, m) {
  serveStyleFile(path.join(this.staticRoot, 'styles', m.id, 'style.json'), m.id, req, res)
}

Api.prototype.stylesGetStatic = function (req, res, m) {
  this.ecstatic(req, res)
}

Api.prototype.stylesGet = function (req, res, m) {
  var self = this
  var ext = m.ext
  var tileid = m.tileid || m.id
  var buf
  var baseFilename = [m.z, m.y, m.x].join(path.sep)
  var asarPath = path.join(self.staticRoot, 'styles', m.id, 'tiles', tileid + '.asar')
  if (ext) buf = asarGet(asarPath, baseFilename + '.' + ext)
  else {
    // Guess the extension
    var guesses = ['png', 'jpg', 'jpeg']
    guesses.every((_ext) => {
      var filename = baseFilename + '.' + _ext
      buf = asarGet(asarPath, filename)
      if (buf) {
        // if the file exists in the asar, lets use that ext
        ext = _ext
        return false
      }
      return true
    })
  }

  // if theres a buffer then the file exists and we know the extension
  if (buf) {
    var mimeType = mime.getType(ext)
    console.log(ext, mimeType)
    if (mimeType) res.setHeader('content-type', mimeType)

    // Set gzip encoding on {mvt,pbf} tiles.
    if (/mvt|pbf$/.test(ext)) res.setHeader('content-encoding', 'gzip')

    res.end(buf)
  } else {
    return handleError(res, errors.NotFound())
  }
}

Api.prototype.syncDestroy = function (req, res, m) {
  this.core.sync.close(function () {
    res.end()
  })
}

Api.prototype.syncJoin = function (req, res, m, q) {
  try {
    if (q.name) this.core.sync.setName(q.name)
    this.core.sync.join(q.project_key)
    res.end()
  } catch (err) {
    handleError(res, err)
  }
}

Api.prototype.syncLeave = function (req, res, m, q) {
  try {
    this.core.sync.leave(q.project_key)
    res.end()
  } catch (err) {
    handleError(res, err)
  }
}

Api.prototype.syncListen = function (req, res, m) {
  this.core.sync.listen(function () {
    res.end()
  })
}

Api.prototype._peers = function () {
  return this.core.sync.peers()
    .map(function (peer) {
      var res = Object.assign({}, peer)
      delete res.connection
      return res
    })
}

Api.prototype.syncPeers = function (req, res, m, q) {
  var self = this
  res.setHeader('Content-Type', 'application/json')
  if (!q.interval) {
    send(res, 'peers', self._peers())
    return res.end()
  }

  var closed = false
  var interval = setInterval(function () {
    if (closed) return
    send(res, 'peers', self._peers())
  }, q.interval)

  res.once('error', done)
  res.once('close', done)

  function done () {
    clearInterval(interval)
    closed = true
    res.end()
  }
}

Api.prototype.syncStart = function (req, res, m, q) {
  var self = this
  try {
    var events = this.core.sync.replicate(q, self.opts)
  } catch (err) {
    return onerror(res, err)
  }

  if (!events) return onerror(res, 'Target not found')

  var debounceProgress = debounce(onprogress, q.interval || 2000)
  events.on('progress', debounceProgress)
  events.once('error', onend)
  events.once('end', onend)

  send(res, 'replication-started')

  function onprogress (progress) {
    send(res, 'replication-progress', progress)
  }

  function onend (err) {
    debounceProgress.clear()
    events.removeListener('progress', debounceProgress)
    if (err) return onerror(res, err.message)
    send(res, 'replication-complete')
    res.end()
  }

  function onerror (res, err) {
    res.statusCode = 500
    var str = JSON.stringify({ topic: 'replication-error', message: (err && err.message) || err }) + '\n'
    res.end(str)
  }
}

Api.prototype.close = function (cb) {
  this.core.sync.close(cb)
}

function send (res, topic, message) {
  var str = JSON.stringify({ topic, message }) + '\n'
  res.write(str)
}

function asarGet (archive, fn) {
  try {
    return asar.extractFile(archive, fn)
  } catch (e) {
    return undefined
  }
}

function serveStyleFile (styleFile, id, req, res) {
  fs.stat(styleFile, function (err, stat) {
    if (err) {
      res.statusCode = 500
      res.end(err.toString())
      return
    }
    fs.readFile(styleFile, 'utf8', function (err, data) {
      if (err) {
        res.statusCode = 500
        res.end(err.toString())
        return
      }
      data = Buffer.from(data.replace(/\{host\}/gm, 'http://' + req.headers.host + '/styles/' + id))
      res.setHeader('content-type', 'application/json; charset=utf-8')
      res.setHeader('last-modified', (new Date(stat.mtime)).toUTCString())
      res.setHeader('cache-control', 'max-age=' + 5 * 60) // 5 minutes
      res.setHeader('content-length', data.length)
      res.setHeader('Access-Control-Allow-Headers', 'Authorization, Content-Type, If-Match, If-Modified-Since, If-None-Match, If-Unmodified-Since')
      res.setHeader('Access-Control-Allow-Origin', '*')
      res.statusCode = 200
      res.end(data)
    })
  })
}
